addons/hr_holidays/models/hr_leave_type.py

```diff
@@ -602,3 +345,209 @@ def action_see_accrual_plans(self):
             'default_time_off_type_id': self.id,
         }
         return action
+
+    # ------------------------------------------------------------
+    # Leave - Allocation link methods
+    # ------------------------------------------------------------
+
+    @api.model
+    def get_allocation_data_request(self, date=None):
+        self = self.search([
+            '|',
+            ('company_id', 'in', self.env.context.get('allowed_company_ids')),
+            ('company_id', '=', False),
+        ])
+        employee = self.env['hr.employee'].browse(self._get_contextual_employee_id())
+        return self.get_allocation_data(employee, date)[employee]
+
+    def get_allocation_data(self, employees, date=None):
+        allocation_data = {emp: [] for emp in employees}
+        if date and isinstance(date, str):
+            date = datetime.fromisoformat(date).date()
+        elif date and isinstance(date, datetime):
+            date = date.date()
+        elif not date:
+            date = fields.Date.today()
+
+        allocations_leaves_consumed = self.get_leaves_consumed(employees)
+        leave_type_with_req_allocation = self.filtered(lambda lt: lt.requires_allocation == 'yes')
+
+        for employee in employees:
+            for leave_type in leave_type_with_req_allocation:
+                if len(allocations_leaves_consumed[employee][leave_type].items()) == 0:
+                    continue
+                lt_info = (
+                    leave_type.name,
+                    {
+                        'remaining_leaves': 0,
+                        'virtual_remaining_leaves': 0,
+                        'max_leaves': 0,
+                        'accrual_bonus': 0,
+                        'leaves_taken': 0,
+                        'virtual_leaves_taken': 0,
+                        'leaves_requested': 0,
+                        'leaves_approved': 0,
+                        'closest_allocation_remaining': 0,
+                        'closest_allocation_expire': False,
+                        'request_unit': leave_type.request_unit,
+                        'icon': leave_type.sudo().icon_id.url,
+                    },
+                    leave_type.requires_allocation,
+                    leave_type.id)
+                allocations_with_remaining_leaves = self.env['hr.leave.allocation']
+                for allocation, data in allocations_leaves_consumed[employee][leave_type].items():
+                    # We only need the allocation that are valid at the given date
+                    if allocation:
+                        if allocation.date_from > date:
+                            continue
+                        if allocation.date_to and allocation.date_to < date:
+                            continue
```

Comment:
could be one if statment

Reply:
I prefer to leave it like this to separate clearly each condition to go out of the loop :)